device-configuration

gisela-ibro opened the pull request.

gisela-ibro committed:
Device Configuration Editor and Verification of Missing Information and Errors in Device Configuration, Guided by Tests.

gisela-ibro committed:
Device Configuration Editor and Verification of Missing Information and Errors in Device Configuration, Guided by Tests.

gisela-ibro committed:
Device Configuration Editor and Verification of Missing Information and Errors in Device Configuration, Guided by Tests.

gisela-ibro committed:
Device Configuration Editor and Verification of Missing Information and Errors in Device Configuration, Guided by Tests.

gisela-ibro committed:
Device Configuration Editor and Verification of Missing Information and Errors in Device Configuration, Guided by Tests.

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfiguration.java:
```diff
@@ -0,0 +1,20 @@
+            e.printStackTrace();
+            return "{}";
+        }
+    }
```
toJson() is unnecessary in that class, we want Domain Objects / Value Objects to be as technology free as possible.
Containing just business logic.
Technical code should be added in Adapter where we need present that value as JSON for REST of document store.

if you just want to verify state of such objects in tests, you can always use:
Assertions.assertThat(editor.toDeviceConfiguratoin().ownership()).isEqualTo(someOwnership());

or as we ware using JsonAssert.assertThat(editor);


But it is nice that you found how to turn object into JSON with ObjectMapper, look on class JsonConfiguration has similar logic 👍🏻

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void resetToDefault() {
+        this.openingHours=OpeningHours.alwaysOpened();
```
lets add if(!Objects.equals(this.field,newValue)) like in assignToOwner, then we Weill not emit same event twice in case if we will process same request twice.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void setSettings(Settings settings) {
+        this.settings=this.settings.merge(settings);
```
lets add if(!Objects.equals(this.field,newValue)) like in assignToOwner, then we Weill not emit same event twice in case if we will process same request twice.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void setLocation(Location location) {
+        this.location=location;
```
lets add if(!Objects.equals(this.field,newValue)) like in assignToOwner, then we Weill not emit same event twice in case if we will process same request twice.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,78 @@
+        events.add(new DomainEvent.LocationChanged(deviceId, location));
+    }
+    public void changeOpeningHours(OpeningHours openingHours) {
+            this.openingHours = openingHours;
```
lets add if(!Objects.equals(this.field,newValue)) like in assignToOwner, then we Weill not emit same event twice in case if we will process same request twice.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,78 @@
+            events.add(new DomainEvent.OpeningHoursChanged(deviceId, openingHours));
+    }
+
+    public void uninstallDevice() {
```
AWESOME, it was on EventStorming diagrams, but we didn't mentioned that command at all.
Perfectly spotted, well implemented.
🔥❤️

michal-michaluk commented on src/test/java/device/DeviceConfigurationEditorTest.java:
```diff
@@ -0,0 +1,320 @@
+        // given
+        DeviceConfigurationEditor editor = DeviceConfigurationEditor.createNewDevice(deviceId);
+        // when
+        String expectedJson = """
```
if you will use `JsonAssert.assertThat().isExactlyLike(""" { ... json content ... } """)`
then syntax hilighting will be applied to json, and you don't need to care about whitespaces.

If you wonder how to suggest to IntelliJ that syntax highlight try: `@Language("JSON") String expectedJson = """`


but please change current solution to:

- Option A: less work, more mess in test
  `JsonAssert.assertThat().isExactlyLike(""" { ... } """)`
- Option B: a bit more work, more clean tests
  ```java
  var config = editor.toDeviceConfiguration()
  Assertions.assertThat(config.ownership()).isEqualTo(Ownership.unowned());
  ```
- Option C: far more work, but sexy tests
   you can create class dedicated to verify sate of configuration, lets call it `DeviceConfigurationAssert`
   static method to get instance:
   `DeviceConfigurationAssert.assertThat(DeviceConfigurationEditor editor)`
   `DeviceConfigurationAssert.assertThat(DeviceConfiguration config)`
   internally that class can thave single field `private final DeviceConfiguration actual;`
   and buch of methods like hasOwnership(Ownership expected), all such methods should `return this;` so type returned form method is DeviceConfigurationAssert.
   in such case we can do
   ```java
  assertThat(editor)
    .hasOwnership(Ownership.unowned())
    .hasNoLocation()
    .hasOpeningHours(OpeningHours.alwaysOpen())
  ```
  you can read a bit more https://mikulskibartosz.name/assert-object-pattern
  
  

**NOTE:** Same comment apply to all current tests.

michal-michaluk commented on src/test/java/device/DeviceConfigurationEditorTest.java:
```diff
@@ -0,0 +1,320 @@
+        editor.assignToOwner(someOwnership());
+        editor.setLocation(someLocationInCity());
+        // when
+        editor.assignToOwner(Ownership.unowned());
```
AWESOME: Great test scenario ! 👌🏼

gisela-ibro committed:
Device Coniguration Assert Class added and Device Configuration Editor Test modified.

gisela-ibro committed:
Device Coniguration Assert Class added and Device Configuration Editor Test modified.

gisela-ibro committed:
Merge remote-tracking branch 'origin/device-configuration' into device-configuration

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

michal-michaluk commented on build.gradle:
```diff
@@ -58,6 +58,9 @@ dependencies {
     testImplementation "au.com.dius.pact.provider:junit5spring:4.3.16"
     testImplementation "org.skyscreamer:jsonassert:1.5.3"
     testImplementation "com.tngtech.archunit:archunit-junit5:1.3.0"
+    testImplementation 'org.assertj:assertj-core:3.24.2'
```
Ok so dependency was not there...
Good job, please only change ' into "
No big difference, just keep it like others

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void resetToDefault() {
+        if (!Objects.equals(this.openingHours, OpeningHours.alwaysOpened())) {
+            this.openingHours = OpeningHours.alwaysOpened();
+        }
+        if (!Objects.equals(this.location, null)) {
+            setLocation(null);
+        }
+        if (!Objects.equals(this.settings, Settings.defaultSettings())) {
+            setSettings(Settings.defaultSettings());
+        }
```
If-s are already in underlying setLocation, setOwnership methods. So in reset methods we don't need to duplicate them. 

Please add method setOpeningHours if not exists, and use it here.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void setSettings(Settings settings) {
+        if (!Objects.equals(this.settings, settings)) {
+            this.settings = this.settings.merge(settings);
```
That is a bit tricky in Settings case:

You merge input with current Settings. 

You need to compare that result with current Settings.

If there is a difference then you perform change and emit event.

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void uninstallDevice() {
+        this.ownership = Ownership.unowned();
```
Let's use assignToOwner method, to emit events it contains already resetToDefault.
So we want only emit additional DeviceRemoved

michal-michaluk commented on src/test/java/device/DeviceConfigurationAssert.java:
```diff
@@ -0,0 +1,44 @@
+        Assertions.assertThat(actual.violations().showOnMapButNoPublicAccess()).isEqualTo(showOnMapButNoPublicAccess);
+        return this;
+    }
+    public DeviceConfigurationAssert hasNoViolations() {
```
Very convenient 👍

michal-michaluk commented on src/test/java/device/DeviceConfigurationAssert.java:
```diff
@@ -0,0 +1,44 @@
+        Assertions.assertThat(actual.settings()).isEqualTo(expected);
+        return this;
+    }
+    public DeviceConfigurationAssert hasViolations(boolean operatorNotAssigned,
+                                                   boolean providerNotAssigned,
+                                                   boolean locationMissing,
+                                                   boolean showOnMapButMissingLocation,
+                                                   boolean showOnMapButNoPublicAccess) {
```
I can imagine separate methods for each. Methods with multiple params of same type are error prone and hard to use.

Consider making something fancy like: `.hasOnlyViolations(Violations.builder().locationMissing(true))`

michal-michaluk commented on src/test/java/device/DeviceConfigurationEditorTest.java:
```diff
@@ -0,0 +1,109 @@
+        DeviceConfigurationEditor editor = DeviceConfigurationEditor.createNewDevice(deviceId);
+        Ownership ownership = someOwnership();
+        editor.assignToOwner(ownership);
+        DeviceConfigurationAssert.assertThat(editor)
+                .hasOwnership(ownership)
+                .hasLocation(null)
+                .hasOpeningHours(OpeningHours.alwaysOpened())
+                .hasSettings(Settings.defaultSettings())
```
You may consider checking only what is important in given scenario.

Assert object makes it very compact, so it looks really slick even when checking all fields.

But try only relevant checks for each test scenario and take your decision.

michal-michaluk reviewed (changes requested):
Very good improvement.
As reward next round of suggestions.

And general comment:
please reformat ctrl+alt+l on src/ folder.
And please enable commit actions in IntelliJ to reformat during commit.

gisela-ibro committed:
Device Coniguration Assert Class Updated and Tests for Domain Event Added

gisela-ibro committed:
Merge remote-tracking branch 'origin/device-configuration' into device-configuration

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

michal-michaluk commented on src/main/java/devices/configuration/device/DeviceConfigurationEditor.java:
```diff
@@ -0,0 +1,88 @@
+@AllArgsConstructor
+public class DeviceConfigurationEditor {
+    private final String deviceId;
+    List<DomainEvent>events;
```
Can be final. Scope is ok 👍

michal-michaluk commented on src/test/java/device/DeviceConfigurationEditorTest.java:
```diff
@@ -0,0 +1,160 @@
+        List<DomainEvent> expectedEvents = List.of(new DomainEvent.OwnershipUpdated(deviceId, ownership));
+        DeviceConfigurationAssert.assertThat(editor)
+                .hasOwnership(ownership)
+                .hasEvents(expectedEvents)
```
Probably variable expectedEvents can be inlined ctrl+alt+n.

You can express meaning in method name, ex. `.emmitedEvents(List.of(...))`

Consider adding another method with "varargs" signature:

`emmitedEvents(DomainEvent... expected)`

michal-michaluk commented on src/test/java/device/DeviceConfigurationEditorTest.java:
```diff
@@ -0,0 +1,160 @@
+                .hasLocation(null)
+                .hasOpeningHours(OpeningHours.alwaysOpened())
+                .hasSettings(Settings.defaultSettings())
+                .hasEvents(expectedEvents)
```
Vararg will look nice I think

michal-michaluk commented on src/test/java/device/DeviceTestFixture.java:
```diff
@@ -0,0 +1,33 @@
+                )
+        );
+    }
+    public static Settings.SettingsBuilder someSettings() {
```
Can it be moved to Fixture?

Commit:
Fix formatting issues and maintain consistent code style

Commit:
Refactor package structure for device test classes

Commit:
Refactor DeviceConfigurationAssert for cleaner event assertions

Commit:
Update tests to include `DeviceCreated` event assertion

Previously, the tests did not account for the `DeviceCreated` event being triggered before other events. This change ensures the `DeviceCreated` event is always asserted first, improving test accuracy and alignment with the event lifecycle.

Commit:
Extract event assertion logic into a dedicated helper class.

Commit:
Refactor test fixture to separate device editor setup

 Introduction of DeviceConfigurationFixture.given() resetting emitted domain events after setup, leaving clear boundary between 'given' and 'when' sections of test.
 Allow to check only events emitted in
 'when' section

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

Commit:
Refactor variable naming in setSettings for clarity.

Commit:
Remove redundant blank lines in build.gradle

Commit:
Make events list final in DeviceConfigurationEditor.

Commit:
maximize test coverage

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge
